flask/src/main.py

```diff
@@ -191,17 +216,50 @@ def products():
         sentry_sdk.capture_exception(err)
         raise (err)
 
-    try:
-        with sentry_sdk.start_span(op="/api_request", description="function"):
-            headers = parseHeaders(RUBY_CUSTOM_HEADERS, request.headers)
-            r = requests.get(RUBY_BACKEND + "/api", headers=headers)
-            r.raise_for_status()  # returns an HTTPError object if an error has occurred during the process
-    except Exception as err:
-        sentry_sdk.capture_exception(err)
+
+    print("cache_key: " + str(cache_key))
+    get_api_request(cache_key, ruby_delay_time)
 
     return rows
 
 
+def get_api_request(key, delay):
+    start_time = time.time()
+    with sentry_sdk.start_span(op="ruby.api.cache") as span:
+      cached_response = redis_client.get("ruby.api.cache:" + str(key))
+      span.set_data("description", "ruby.api.request")
+      span.set_data("cache.key", key)
+
+      if cached_response is not None:
+          print("> cache hit: " + str(key))
+
+          span.set_data("cache.hit", True)
+          return cached_response
+      else:
+          print("> cache miss: " + str(key))
+          span.set_data("cache.hit", False)
+
+      try:
+          with sentry_sdk.start_span(op="/api_request", description="function"):
+              headers = parseHeaders(RUBY_CUSTOM_HEADERS, request.headers)
+              r = requests.get(RUBY_BACKEND + "/api", headers=headers)
+              r.raise_for_status()  # returns an HTTPError object if an error has occurred during the process
+              # For demo show we want to show cache misses so only save 1 / 100
+              if key == 7:
+                redis_client.set("ruby.api.cache:" + str(key), key)
+      except Exception as err:
+          sentry_sdk.capture_exception(err)
+
+      time_delta = time.time() - start_time
+      sleep_time = delay - time_delta
+      if time_delta > 0:
+        time.sleep(sleep_time)
```

Comment:
Is there a reason not to move this under inside `with`?
```
with sentry_sdk.start_span(op="/api_request"
```
<img width="1158" alt="Screenshot 2024-11-07 at 11 24 51 AM" src="https://github.com/user-attachments/assets/09e20b10-190c-4fea-a80d-05d05272dfd6">

Reply:
Nope, good catch.